rollup/circuitcapacitychecker/libzkp/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+#![feature(once_cell)]
+
+pub mod checker {
+    use crate::utils::{c_char_to_vec, vec_to_c_char};
+    use libc::c_char;
+    use prover::zkevm::{CircuitCapacityChecker, RowUsage};
+    use serde_derive::{Deserialize, Serialize};
+    use std::cell::OnceCell;
+    use std::collections::HashMap;
+    use std::panic;
+    use std::ptr::null;
+    use types::eth::BlockTrace;
+
+    #[derive(Debug, Clone, Deserialize, Serialize)]
+    pub struct RowUsageResult {
+        pub acc_row_usage: Option<RowUsage>,
+        pub tx_row_usage: Option<RowUsage>,
+        pub error: Option<String>,
+    }
+
+    static mut CHECKERS: OnceCell<HashMap<u64, CircuitCapacityChecker>> = OnceCell::new();
+
+    /// # Safety
+    #[no_mangle]
+    pub unsafe extern "C" fn init() {
+        env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("debug"))
+            .format_timestamp_millis()
+            .init();
+        let checkers = HashMap::new();
+        CHECKERS.set(checkers).unwrap();
+    }
+
+    /// # Safety
+    #[no_mangle]
+    pub unsafe extern "C" fn new_circuit_capacity_checker() -> u64 {
+        let id = CHECKERS.get_mut().unwrap().len() as u64;
+        let checker = CircuitCapacityChecker::new();
+        CHECKERS.get_mut().unwrap().insert(id, checker);
```

Comment:
```suggestion
        let mut checkers = CHECKERS.get_mut().expect("circuit capacity checker used before initialization");
        let id = checkers.len() as u64;
        let checker = CircuitCapacityChecker::new();
        checkers.insert(id, checker);
```

Also, should we do any concurrency control here? Two concurrent calls to `new_circuit_capacity_checker` might end up using the same id.

Reply:
fixed in https://github.com/scroll-tech/go-ethereum/pull/321/commits/baf2eccf12d9bcf16a0f63ed7beeff7a7169f162 & https://github.com/scroll-tech/go-ethereum/pull/321/commits/714afd721af62aa19b2725994f8154509950f629